### PR TITLE
Limit memory usage when capturing SOAP request bodies

### PR DESCRIPTION
**Pull request by gregkalapos**
Limit memory usage by creating an ArrayPool with limited MaxArrayLength and skipping capturing SOAP bodies larger than that.

### CODE REVIEW

File: src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs [hunk 1]
```diff
@@ -20,6 +20,8 @@ internal static class SoapRequest
 		private const string SoapActionHeaderName = "SOAPAction";
 		private const string ContentTypeHeaderName = "Content-Type";
 		private const string SoapAction12ContentType = "application/soap+xml";
+		private const int MaxRequestBodySizeInBytesToCapture = 1024 * 10;
```
Comment (gregkalapos) [inline src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs:23]: This is a fairly arbitrary limit and I'm totally open to change it if we have any better number. I just naively think that capturing something bigger than this doesn't make a lot of sense, so I choose this limit.

File: src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs [hunk 2]
```diff
@@ -20,6 +20,8 @@ internal static class SoapRequest
...
+		private static readonly ArrayPool<byte> ArrayPool = ArrayPool<byte>.Create(MaxRequestBodySizeInBytesToCapture, 50);
```
Comment (gregkalapos) [inline src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs:24]: `maxArraysPerBucket = 50` is [the default](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/Buffers/ConfigurableArrayPool.cs#L15C63-L15C65).

**Comment by gregkalapos** [comment]
I see testcontainer setup issues, seems unrelated:

```
2m 6s
Run atomicjar/testcontainers-cloud-setup-action@v1
Run # download and run the agent at the background
Downloading https://app.testcontainers.cloud/download/testcontainers-cloud-agent_latest_windows_x86-64.exe
2024-02-01T19:28:47.858Z --- INF Testcontainers Cloud Agent arch=amd64 os="Windows Server 2022 Datacenter 10.0 (6.3)" version=1.7.0.4660+eac8767
2024-02-01T19:28:47.927Z --- INF Initializing config cloud="C:\\Users\\runneradmin\\.config\\testcontainers\\cloud.properties" dir="C:\\Users\\runneradmin\\.config\\testcontainers" tc="C:\\Users\\runneradmin\\.testcontainers.properties"
2024-02-01T19:28:47.928Z --- WRN Failed to parse TC_CLOUD_TOKEN as token
2024-02-01T19:28:47.929Z --- INF Creating new state file location="C:\\Users\\runneradmin\\AppData\\Local\\AtomicJar\\testcontainers.cloud.desktop\\state.json"
2024-02-01T19:28:47.930Z --- INF Assigning new installation ID iid=iid_10ac83f2-34d2-44d8-9f0e-8c2810d7b7aa
ERROR: incorrect token
Run # waiting for agent successful connection
2024-02-01T19:28:48.008Z --- INF Testcontainers Cloud Agent arch=amd64 os="Windows Server 2022 Datacenter 10.0 (6.3)" version=1.7.0.4660+eac8767
2024-02-01T19:28:48.028Z --- INF Initializing config cloud="C:\\Users\\runneradmin\\.config\\testcontainers\\cloud.properties" dir="C:\\Users\\runneradmin\\.config\\testcontainers" tc="C:\\Users\\runneradmin\\.testcontainers.properties"
2024-02-01T19:28:48.028Z --- INF Waiting up to 120 seconds for the agent to start...
2024-02-01T19:30:48.041Z --- FTL Timed out waiting for agent to be running addr=unknown
Error: Process completed with exit code 1.
```

File: src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs [hunk 3]
```diff
@@ -102,20 +104,22 @@ internal static string GetSoap12ActionFromHttpRequest(IApmLogger logger, HttpReq
...
 		internal static string GetSoap12ActionFromInputStream(IApmLogger logger, Stream stream)
 		{
-			var shared = ArrayPool<byte>.Shared;
-			var bytes = shared.Rent((int)stream.Length);
+			if (stream.Length > MaxRequestBodySizeInBytesToCapture)
+			{
+				logger.Info()
+					?.Log(
+						$"SOAP request body exceeds {MaxRequestBodySizeInBytesToCapture} bytes - skipping request body capturing to limit memory usage");
+				return null;
```
Comment (stevejgordon) [inline src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs:112]: Is there any value in capturing the initial bytes, as much as we can fit in the buffer? If not, should we skip renting the array at all?

**Comment by Mpdreamz** [comment]
> I see testcontainer setup issues, seems unrelated:
> 
> ```
> 2m 6s
> Run atomicjar/testcontainers-cloud-setup-action@v1
> Run # download and run the agent at the background
> Downloading https://app.testcontainers.cloud/download/testcontainers-cloud-agent_latest_windows_x86-64.exe
> 2024-02-01T19:28:47.858Z --- INF Testcontainers Cloud Agent arch=amd64 os="Windows Server 2022 Datacenter 10.0 (6.3)" version=1.7.0.4660+eac8767
> 2024-02-01T19:28:47.927Z --- INF Initializing config cloud="C:\\Users\\runneradmin\\.config\\testcontainers\\cloud.properties" dir="C:\\Users\\runneradmin\\.config\\testcontainers" tc="C:\\Users\\runneradmin\\.testcontainers.properties"
> 2024-02-01T19:28:47.928Z --- WRN Failed to parse TC_CLOUD_TOKEN as token
> 2024-02-01T19:28:47.929Z --- INF Creating new state file location="C:\\Users\\runneradmin\\AppData\\Local\\AtomicJar\\testcontainers.cloud.desktop\\state.json"
> 2024-02-01T19:28:47.930Z --- INF Assigning new installation ID iid=iid_10ac83f2-34d2-44d8-9f0e-8c2810d7b7aa
> ERROR: incorrect token
> Run # waiting for agent successful connection
> 2024-02-01T19:28:48.008Z --- INF Testcontainers Cloud Agent arch=amd64 os="Windows Server 2022 Datacenter 10.0 (6.3)" version=1.7.0.4660+eac8767
> 2024-02-01T19:28:48.028Z --- INF Initializing config cloud="C:\\Users\\runneradmin\\.config\\testcontainers\\cloud.properties" dir="C:\\Users\\runneradmin\\.config\\testcontainers" tc="C:\\Users\\runneradmin\\.testcontainers.properties"
> 2024-02-01T19:28:48.028Z --- INF Waiting up to 120 seconds for the agent to start...
> 2024-02-01T19:30:48.041Z --- FTL Timed out waiting for agent to be running addr=unknown
> Error: Process completed with exit code 1.
> ```

Yeah this is due to forks not having access to secrets, rightfully so.

https://github.com/elastic/apm-agent-dotnet/pull/2273 addresses this.

https://github.com/elastic/apm-agent-dotnet/pull/2275 is the continuation of that to ignore certain tests if run on from a pr from a fork that needs TestContainers

File: src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs (hunk 1 above)
Comment (Mpdreamz) [reply src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs:23]: 10kb seems reasonable, even if it is SOAP 🙈

**Review by Mpdreamz — CHANGES_REQUESTED** [review]
https://github.com/elastic/apm-agent-dotnet/pull/2274/files#r1475983104

File: src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs (hunk 3 above)
Comment (gregkalapos) [reply src/integrations/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs]: Yeah, that makes sense. I adapted the code to always capture up to `MaxRequestBodySizeInBytesToCapture`.

**Review by Mpdreamz — APPROVED** [review]
Lgtm!

**Pull request merged**